src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadExerciseResource.java

```diff
@@ -106,6 +110,11 @@ public FileUploadExerciseResource(FileUploadExerciseService fileUploadExerciseSe
 
         FileUploadExercise result = fileUploadExerciseRepository.save(fileUploadExercise);
 
+        // Generate achievements if enabled in course and exercise not part of exam
+        if (course.getHasAchievements() && (result.getExerciseGroup() == null || result.getExerciseGroup().getExam() == null)) {
```

Comment:
If I see it correctly, this is used for all exercise types. To reduce code duplication you could extract this into `ExerciseService.java` :)

Reply:
yeah, makes sense :) thanks